docs/execution.md

```diff
@@ -0,0 +1,208 @@
+# Smart contracts
+
+## Smart Contract Engine
+
+### Rationale
+
+The smart contract engine (SCE) is decoupled from the consensus system (CSS); they maintain separate ledgers: the SCE ledger and the CSS ledger.
+
+The CSS informs the SCE whenever a block becomes CSS-final, and whenever the blockclique changes in any way.
+
+The SCE executes the blockclique operations after sorting them by increasing block slot: this provides an absolute order of execution.
+
+A special feedback loop is required to send coins from the SCE to the CSS.
+
+A unit of "gas" represents a unit of computational cost. Different instructions might cost different amounts of gas.
+
+### The ExecuteSC operation
+
+The following new operation type is defined:
+
+```rust
+OperationType::ExecuteSC {
+    max_gas: u32,
+    gas_price: Amount,
+    coins: Amount,
+    bytecode: Vec<u8>,
+}
+```
+
+Where:
+* `bytecode` is smart contract bytecode that is understood by the SCE but not the CSS
```

Comment:
Do we enforce it to be wasm byte code ? If not maybe add a language field that could be an enum representing all supported languages (that could have just one variant for now)

Reply:
The idea is that the CSS side does not interpret this field at all. It is only interpreted in the SCE.

For now we went for having only one type of bytecode: WASM, but for sure we could add a field describing which bytecode (and version!) it is